packages/controllers/src/permissions/PermissionController.ts

```diff
@@ -0,0 +1,1884 @@
+};
+
+/**
+ * Clears all permissions from the {@link PermissionController}.
+ */
+export type ClearPermissions = {
+  type: `${typeof controllerName}:clearPermissions`;
+  handler: () => void;
+};
+
+// TODO: Implement all desired actions
+/**
+ * The {@link ControllerMessenger} actions of the {@link PermissionController}.
+ */
+export type PermissionControllerActions =
+  | ClearPermissions
+  | GetPermissionControllerState
+  | GetSubjects
+  | HasPermissions;
+
+/**
+ * The generic state change event of the {@link PermissionController}.
+ */
+export type PermissionControllerStateChange = {
+  type: `${typeof controllerName}:stateChange`;
+  payload: [PermissionControllerState<GenericPermission>, Patch[]];
+};
+
+/**
+ * The {@link ControllerMessenger} events of the {@link PermissionController}.
+ *
+ * The permission controller only emits its generic state change events.
+ * Consumers should use selector subscriptions to subscribe to relevant
+ * substate.
+ */
+export type PermissionControllerEvents = PermissionControllerStateChange;
+
+/**
+ * The external {@link ControllerMessenger} actions available to the
+ * {@link PermissionController}.
+ */
+type AllowedActions =
+  | AddApprovalRequest
+  | HasApprovalRequest
+  | AcceptApprovalRequest
+  | RejectApprovalRequest;
+
+/**
+ * The messenger of the {@link PermissionController}.
+ */
+export type PermissionControllerMessenger = RestrictedControllerMessenger<
+  typeof controllerName,
+  PermissionControllerActions | AllowedActions,
+  PermissionControllerEvents,
+  AllowedActions['type'],
+  never
+>;
+
+export type GenericPermissionController = PermissionController<
+  PermissionSpecificationBase<string>,
+  CaveatSpecificationBase<string>
+>;
+
+/**
+ * Options for {@link PermissionController.grantPermissions}.
+ */
+type GrantPermissionsOptions = {
+  approvedPermissions: RequestedPermissions;
+  subject: PermissionSubjectMetadata;
+  preserveExistingPermissions?: boolean;
+  requestData?: Record<string, unknown>;
+};
+
+/**
+ * Describes the possible results of a {@link CaveatMutator} function.
+ */
+export enum CaveatMutatorOperation {
+  noop,
+  updateValue,
+  deleteCaveat,
+  revokePermission,
+}
+
+/**
+ * Given a caveat value, returns a {@link CaveatMutatorOperation} and, optionally,
+ * a new caveat value.
+ *
+ * @see {@link PermissionController.updatePermissionsByCaveat} for more details.
+ * @template Caveat - The caveat type for which this mutator is intended.
+ * @param caveatValue - The existing value of the caveat being mutated.
+ * @returns A tuple of the mutation result and, optionally, the new caveat
+ * value.
+ */
+export type CaveatMutator<Caveat extends GenericCaveat> = (
+  caveatValue: Caveat['value'],
+) =>
+  | [Exclude<CaveatMutatorOperation, CaveatMutatorOperation.updateValue>]
+  | [CaveatMutatorOperation.updateValue, GenericCaveat['value']];
+
+/**
+ * Extracts the permission type or types from the specified permission and
```

Comment:
This makes it sound like it's extracting just the types. But it's getting the whole object. Which was intended?

Reply:
Fixed in ea899da